Java to Kotlin conversion

ashwinkey04 opened the pull request:
Solves Issue(s):  #1530 

Converted the following classes which were previously in Java, to Kotlin.

1. Passcode activity
2. Constants
3. Convert Registration activity
4. Home activity and fragments
5. Base activity and fragments
6. API related classes 
7. All View classes
8. All classes under `/main/../ui` (Activities, Adapters, Enums, Views, Widgets, Fragments)
9. All Adapter classes
10. All Presenter classes
11. All JUnit tests 
12. All activity classes
13. All model classes
14. All Injection classes
15. MifosSelfService class
16. commonTest and androidTest classes 

---

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them 
(Separate commits for easier reviews. Will squash after the final review)

ashwinkey04 committed:
Convert Passcode activity and constants to Kotlin

ashwinkey04 committed:
Convert Registration activity to Kotlin

ashwinkey04 committed:
Convert Home activity and fragment to Kotlin

ashwinkey04 committed:
Convert Base activity and fragment to Kotlin

ashwinkey04 committed:
Convert API related classes to Kotlin

ashwinkey04 committed:
Convert all Java views to Kotlin

ashwinkey04 committed:
feat: Convert all fragments to Kotlin

ashwinkey04 committed:
feat: Convert enum and widget classes to Kotlin

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/AboutUsFragment.kt:
```diff
@@ -0,0 +1,60 @@
+        rootView = inflater.inflate(R.layout.fragment_about_us, container, false)
+        ButterKnife.bind(this, rootView!!)
+        setToolbarTitle(getString(R.string.about_us))
+        tvAppVersion!!.text = getString(R.string.app_version, BuildConfig.VERSION_NAME)
```
I think it is much better to use the safe call `?.` (One of the biggest advantages of using Kotlin) instead of `!!`, because it leads the App to crash if our view or any other item is null or it has not been initiated. You can replace all `!!` to `?.` in a file, simply by entering `Ctlr+r`.

ashwinkey04 replied:
Thank you for your insights @jawidMuhammadi. I guess I had failed to notice a lot of not-null assertion operators since this conversion was straight from Java. Though there are quite a few functions in the app which requires a non-null object (For instance, `ButterKnife.bind()` is a Java function and doesn't support null safety yet), I would check for usages where safe call is possible and update it soon. Thanks again :)

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/AboutUsFragment.kt:
```diff
@@ -0,0 +1,60 @@
+*/   class AboutUsFragment : BaseFragment() {
+    @kotlin.jvm.JvmField
+    @BindView(R.id.tv_app_version)
+    var tvAppVersion: TextView? = null
```
If we are using **Kotlin** then isn't it better to optimize our code a little more and decrease lines of code. Like, here, I think there is no need for those 3 lines of code(23,24,25) and in all other places. We can simply call `tv_app_version` whenever we want using **Kotlin android-synthetics**

ashwinkey04 replied:
Sorry for the late reply on this, I just noticed this comment. Firstly, KotlinX android synthetics is now regarded to be not a recommended practice anymore [[1]](https://android-review.googlesource.com/c/platform/frameworks/support/+/882241) [[2]](https://old.reddit.com/r/androiddev/comments/ala9p2/why_kotlinx_synthetic_is_no_longer_a_recommended/efdvpkg/) since it was replaced with `findViewById`. I did not use that either since the app already used ButterKnife for ViewBinding and I had to primarily focus on completely converting the app to Kotlin first. I will discuss and get back on if I should replace ButterKnife. Thank you for the suggestions and do let me know if there are any more optimization suggestions from your side :)

ashwinkey04 committed:
feat: Convert all Util classes to Kotlin

ashwinkey04 committed:
feat: Convert all adapter classes to Kotlin

Null check all fragment classes and convert the depended presenter classes

ashwinkey04 committed:
feat: Convert Presenters and JUnit tests to Kotlin

ashwinkey04 committed:
feat: Replace non-null assertions

ashwinkey04 committed:
feat: Convert activities to Kotlin

ashwinkey04 committed:
feat: Convert models to Kotlin

ashwinkey04 committed:
fix: Charges fetching error

fix: Charges User profile fragment bug

ashwinkey04 committed:
fix: Empty spinners and Non nullable OnItemSelected definitions

ashwinkey04 committed:
feat: Convert all injections to Kotlin

ashwinkey04 committed:
feat: Convert MifosSelfServiceApp class to Kotlin

ashwinkey04 committed:
feat: Convert commonTests and androidTest to Kotlin

Finish converting the app to Kotlin

ashwinkey04 committed:
fix: User profile picture drawable

Fix injection error in preferencesHelper. Fix blank user name in nav drawer head.

Fix set:Inject annotations

ashwinkey04 committed:
fix: Non nullable values in models

fix: Restrict rotation in passcode screen to portrait

ashwinkey04 committed:
fix: Redundant object usages

miPlodder commented on app/src/main/AndroidManifest.xml:
```diff
@@ -56,7 +56,8 @@
         <activity android:configChanges="orientation|screenSize"
                   android:launchMode="singleTask"
                   android:name="org.mifos.mobile.ui.activities.PassCodeActivity"
-                  android:windowSoftInputMode="adjustResize"/>
```
Why is this `screenOrientation` added. Please clarify

ashwinkey04 replied:
If the orientation is landscape in the passcode activity, the UI elements go out of bounds. mifos-mobile-cn also had a similar [problem](https://github.com/openMF/mifos-mobile-cn/issues/108) which was solved by restricting the orientation to only portrait.

edcable commented:
@miPlodder is this ready to be merged?

miPlodder commented:
@edcable This PR looks good to me.

@therajanmaurya If possible can you also take a look at it once and share your thoughts?
